examples/microwave-oven-app/microwave-oven-common/src/microwave-oven-device.cpp

```diff
@@ -0,0 +1,228 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "microwave-oven-device.h"
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::OperationalState;
+using namespace chip::app::Clusters::ModeBase;
+using namespace chip::app::Clusters::MicrowaveOvenControl;
+template <typename T>
+using List                 = chip::app::DataModel::List<T>;
+using ModeTagStructType    = chip::app::Clusters::detail::Structs::ModeTagStruct::Type;
+using OperationalStateEnum = chip::app::Clusters::OperationalState::OperationalStateEnum;
+using Status               = Protocols::InteractionModel::Status;
+
+void ExampleMicrowaveOvenDevice::MicrowaveOvenInit()
+{
+    mOperationalStateInstance.SetOperationalState(to_underlying(OperationalStateEnum::kStopped));
+    mOperationalStateInstance.Init();
+    mMicrowaveOvenModeInstance.Init();
+    mMicrowaveOvenControlInstance.Init();
+}
+
+/**
+ * MicrowaveOvenControl cluster
+ */
+Protocols::InteractionModel::Status
+ExampleMicrowaveOvenDevice::HandleSetCookingParametersCallback(Optional<uint8_t> cookMode, uint32_t cookTime, uint8_t powerSetting)
+{
+    // placeholder implementation
+    Status status;
+    uint8_t reqCookMode;
+    if (cookMode.HasValue())
```

Comment:
Shouldn't this be handled by the cluster implementation?  This is spec-required behavior that we can easily handle in the cluster; we should not be pushing this off to the app layer.

Reply:
@liangpy4

Reply:
In particular, we should be able to set it to `MicrowaveOvenMode::ModeTag::kNormal`, no?

Reply:
Our thought is, due to the MicrowaveOvenMode cluster is not like as MicrowaveOvenControl which is defines the class of delegate and instance in the SDK layer,  it is defined in the APP layer only, if we want to get access of mode's value, we should use the value 0(`ModeNormal`) or 1(`ModeDefrost`)  which are defined in the APP layer.

Now we change to get use of `MicrowaveOvenMode::ModeTag::kNormal`, but stiil need to do some processing to get the mode value(i.e. reqCookMode =  to_underlying(MicrowaveOvenMode::ModeTag::kNormal) - 0x4000), but we think it might not be a good idea.

@bzbarsky-apple Could you left some suggestions on this case? Thanks!




> In particular, we should be able to set it to `MicrowaveOvenMode::ModeTag::kNormal`, no?

Reply:
In the spec we found the mode tag only, but not found mode value about "Normal". Do the index of mode tag is equal with mode value?
 
> Is "Normal" not a constant value that is defined in the spec? If it is, what is the problem with setting it? Either I am not getting something, or there's a spec problem here....

Reply:
>  should use the value 0(ModeNormal) or 1(ModeDefrost) which are defined in the APP layer.

Where are those values coming from?

The spec says:

> [8.122] This field SHALL indicate the value to which the CurrentMode attribute of the Microwave Mode cluster should be set. The value of this field SHALL be one from the list of SupportedModes from the Microwave Mode cluster.
>
> [8.123] If this field is missing, the CurrentMode attribute SHALL be set to Normal by the server.

What does "Normal" mean here, exactly?  And what is the "Microwave Mode" cluster?  Should that be "Microwave Oven Mode", and should Normal be an actual value or "pick some mode that has the Normal tag"?  @rbultman

Reply:
It should be the Microwave Oven Mode cluster and Normal is one of 2 derivative-cluster-specific modes.

Reply:
> Where are those values coming from?

Those values are defined in MicrowaveOvenMode manually. 

In [examples/microwave-oven-app/microwave-oven-common/include/microwave-oven-device.h](https://github.com/project-chip/connectedhomeip/pull/30235/files/ad8e1329e55d2fd516e46dba0c3799d781c136d3#diff-f295c870bd4b7c8b291c996aed1ea6ec4b1eb3b6b24a9104f6f8a211b071aa95)

```
constexpr uint8_t ModeNormal  = 0;
constexpr uint8_t ModeDefrost = 1;
```

Reply:
These should have been defined via the XML for the cluster.  MicrowaveOvenMode::ModeTag::kNormal.

See cluster-enums.h.

Reply:
You should not have to define these manually.

Reply:
The meaning of Normal is up to the manufacturer.  The value is defined in the Microwave Oven Mode cluster.

Reply:
Unfortunately the values of Mode are not defined and are MS.  I think the intention was that there would be a correspondence between the Mode and ModeTags, but that is not prescribed in the spec.

Reply:
What the spec says is "set mode to Normal".  That's not actually possible, since there is no value of mode called Normal.  Can we please get the spec clarified so it's implementable @rbultman ?

Reply:
See this Slack DM: https://csamembers.slack.com/archives/C069NQCPJMR/p1702412512396629

In short, Mode is MS while ModeTag is not.  A Normal 'mode' is the SupportMode that has a ModeTag of Normal in its ModeTags list for that mode.

Reply:
Thanks for @rbultman 's comment!

Reply:
@bzbarsky-apple we have updated the MicrowaveOvenControl server & ModeBase server based on the new required, could you help review one more time when you feel free? Thanks.